File: 004: Palindromic Product/LargestPalindromicProduct.py
import math


def main():
    digits = int(input("Enter number of digits for multiplier: "))

    count = 0
    first_value = 0
    second_value = 0
    product = 0
    palindromic_product = 0
    lower_limit = int(math.pow(10, digits - 1))
    upper_limit = int(math.pow(10, digits))

    for i in range(lower_limit, upper_limit):
        for j in range(lower_limit, upper_limit):
            product = i * j
            if isPalindrome(product):
                count += 1
                first_value = i
                second_value = j
                palindromic_product = product

    print(first_value, "*", second_value, "=", palindromic_product)
    print(count)


def isPalindrome(product):
    remain = 0
    reverse_product = 0
    temp = product
    palindrome = False
    while temp != 0:
        remain = temp % 10
        reverse_product = reverse_product * 10 + remain
        temp = int(temp / 10)

    if product == reverse_product:
        palindrome = True

    return palindrome


main()

File: 004: Palindromic Product/LargestPalindromicProduct.java
import java.util.Scanner;

public class LargestPalindromicProduct {
	/*
	 * A palindromic number reads the same both ways. The largest palindrome made
	 * from the product of two 2-digit numbers is 9009 = 91 × 99.
	 * 
	 * Find the largest palindrome made from the product of two 3-digit numbers.
	 */

	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		System.out.print("Enter number of digits for multiplier: ");
		int digits = scanner.nextInt();
		int count = 0;
		int firstValue = 0;
		int secondValue = 0;
		int product = 0;
		int palindromicProduct = 0;
		for (int i = (int) Math.pow(10, digits - 1); i < (int) Math.pow(10, digits); i++) {
			for (int j = (int) Math.pow(10, digits - 1); j < (int) Math.pow(10, digits); j++) {
				product = i * j;
				if (isPalindrome(product)) {
					count++;
					firstValue = i;
					secondValue = j;
					palindromicProduct = product;
				}
			}
		}
		System.out.println(firstValue + " * " + secondValue + " = " + palindromicProduct);
		System.out.println(count);
		scanner.close();
	}

	public static boolean isPalindrome(int prod) {
		int remainder = 0;
		int reversedProd = 0;
		int temp = prod;
		boolean palindrome = false;
		while (temp != 0) {
			remainder = temp % 10;
			reversedProd = reversedProd * 10 + remainder;
			temp /= 10;
		}

		if (prod == reversedProd) {
			palindrome = true;
		}
		return palindrome;
	}
}
